### ISSUE TITLE
test on my own dataset

### ISSUE THREAD

**Issue by Ethean** [yuanjunchai/IKC#7]
Hi, thanks for your job! I want to test the model on a dataset of myself, what should I do?
I tried to change the path in 
> IKC-master/codes/options/test/test_SFTMD.yml,
   IKC-master/codes/options/test/test_Predictor.yml,
   and IKC-master/codes/options/test/test_Corrector.yml
### and got a pretty low average psnr(23.598937) and ssim(0.645178)

> 19-11-15 17:31:03.120 - INFO: ----Average PSNR/SSIM results for div2k100----
	PSNR: 23.598937 dB; SSIM: 0.645178
19-11-15 17:31:03.120 - INFO: ----Y channel, average PSNR/SSIM----
	PSNR_Y: 25.071245 dB; SSIM_Y: 0.682182
>
My input images were the validation set of div2k(801-900) and all of them are blured with an isotropic gaussian kernel with width 2.0, then downsampled with bicubic interpolation. All the pictures were processed with opencv 4.1.0. The samples of input and the SR output are shown blew 
input
![0802x4](https://user-images.githubusercontent.com/46132747/68938123-a7564280-07d8-11ea-8545-eb400089bbbb.png)
output
![0802_7](https://user-images.githubusercontent.com/46132747/68938191-ce147900-07d8-11ea-9c5a-ce571d4daae9.png)
the log information

> 19-11-15 16:08:18.763 - INFO: step:  1, img:0802            - PSNR: 27.946776 dB; SSIM: 0.775806; PSNR_Y: 29.448376 dB; SSIM_Y: 0.818709.
19-11-15 16:08:25.778 - INFO: step:  2, img:0802            - PSNR: 27.992938 dB; SSIM: 0.776571; PSNR_Y: 29.497938 dB; SSIM_Y: 0.819453.
19-11-15 16:08:32.779 - INFO: step:  3, img:0802            - PSNR: 28.035925 dB; SSIM: 0.777074; PSNR_Y: 29.545070 dB; SSIM_Y: 0.820215.
19-11-15 16:08:39.848 - INFO: step:  4, img:0802            - PSNR: 28.079307 dB; SSIM: 0.777602; PSNR_Y: 29.592643 dB; SSIM_Y: 0.820966.
19-11-15 16:08:46.881 - INFO: step:  5, img:0802            - PSNR: 28.124418 dB; SSIM: 0.778161; PSNR_Y: 29.642248 dB; SSIM_Y: 0.821761.
19-11-15 16:08:53.952 - INFO: step:  6, img:0802            - PSNR: 28.174223 dB; SSIM: 0.778820; PSNR_Y: 29.696932 dB; SSIM_Y: 0.822650.
19-11-15 16:09:00.975 - INFO: step:  7, img:0802            - PSNR: 28.228933 dB; SSIM: 0.779559; PSNR_Y: 29.757185 dB; SSIM_Y: 0.823617.
19-11-15 16:09:00.976 - INFO: step:  7, img:0802            - average PSNR: 28.083217 dB; SSIM: 0.777656; PSNR_Y: 29.597199 dB; SSIM_Y: 0.821053.
19-11-15 16:09:00.976 - INFO: step:  7, img:0802            - max PSNR: 28.228933 dB; SSIM: 0.779559; PSNR_Y: 29.757185 dB; SSIM_Y: 0.823617.
>
what did I miss? Thanks in advance!

**Comment by yuanjunchai** [comment]
> and got a pretty low average psnr(23.598937) and ssim(0.645178)
Do you mean you get the result from the IKC model? Maybe you should follow max results rather than average results, for there are 7 steps in one blurred image, therefore, the max result among the following 7 results might be chosen as the final result. Also, you could find both max and average results in /.log, and maybe we should modify the code to make final results look clearer.

**Comment by yuanjunchai** [comment]
Do you mean you get the result from the IKC model? Maybe you should follow max results rather than average results, for there are 7 steps in one blurred image, therefore, the max result among the following 7 results might be chosen as the final result. Also, you could find both max and average results in /.log, and maybe we should modify the code to make final results look clearer.

**Comment by yuanjunchai** [comment]
There are many outputs in one image, we'd better choose max PSNR result and save this image.

**Comment by Ethean** [comment]
@yuanjunchai Thanks for replying me. I tested the IKC model on my own synthesised dataset. I do check the program and be pretty sure that the log records the max psnr and ssim to calculate the final average results. And as shown above, the sample of SR output is the final output of the sr-network in step seven. I guess I must have missed up something, and I will check my dataset path. Again, thank you !

**Comment by yuanjunchai** [comment]
my pleasure!
BTW, to solve the problem, could you tell me how did you blur the dataset?

**Comment by yuanjunchai** [comment]
Since I've tested div2k, BSDS, set14 and set5, all IKC results perform equally even better than paper proposed.

**Comment by Ethean** [comment]
@yuanjunchai I blurred the image with opencv funtion fileter2D
`image = cv.filter2D(image, -1, kernel)`
and the image is the original bicubic_X4 version of the Div2k validation set, and the kernel is systhesised with matlab function mvnpdf
```
mu=[0,0];
[X,Y]=meshgrid(-4:1:4,-4:1:4);
out_path = 'kernels/test_kernels/';
if isdir(out_path)==0
    mkdir(out_path)
end
sigma = [2.0 0 ; 0 2.0];
kernal = mvnpdf([X(:) Y(:)],mu,sigma);
k = reshape(kernal, size(X));
save([out_path, 'anisotropic2.0', '.mat'], 'k')
```
the kernel size is fixed to 9
and I change the path in 

> IKC-master/codes/options/test/test_SFTMD.yml,
IKC-master/codes/options/test/test_Predictor.yml,
IKC-master/codes/options/test/test_Corrector.yml
>
like 

> dataroot_GT:/home/xx/sdb/dataset/test/Div2k_val/HR/
dataroot_LQ: /home/xx/sdb/dataset/test/Div2k_val/blur/isotropic2/
>
and get the results. So it must be the data? Cause they are in different distribution?

**Comment by yuanjunchai** [comment]
Well, I know what happened. Since the IKC model has been trained by input via blur kernel & downsample setting as IKC paper proposed, namely 21*21 isotropic Gaussian kernel, there is a data distribution gap between pre-trained data and dataset you are provided. 
There are two methods to solve your problem: 1. Train your own model given your synthesized data. 2. Use `python codes/scripts/generate_mod_LR_bic.py` to generate data following training input as proposed in IKC paper.

**Comment by Ethean** [comment]
Thank you ! I am trying to train a new model with my own dataset.

**Comment by yuanjunchai** [comment]
My pleasure.
BTW, I recommend training the new model by the process in our training code, for the F model(SFTMD) need the blurred LR and corresponding true kernel map h. 
Of course, other method is OK for training your own models.

**Comment by Ethean** [comment]
Hi, I made a new dataset with the same method as claimed above. I change the kernel to a 21*21 isotropic Gaussian kernel with width 2.0

> mu=[0,0];
[X,Y]=meshgrid(-10:1:10,-10:1:10);
out_path = 'kernels/test_kernels/';
if isdir(out_path)==0
    mkdir(out_path)
end
sigma = [2.0 0 ; 0 2.0];
kernal = mvnpdf([X(:) Y(:)],mu,sigma);
k = reshape(kernal, size(X));
save([out_path, 'isotropic2.0', '.mat'], 'k')
>
and blured image with `image = cv.filter2D(image, -1, kernel)` and tested on the pretrained model. Samely, I still got a low psnr(24.6) and ssim(0.659). And I am a rookie in blind sr, so I want know how  you blur images in the train set? I did not figure out from the code.Thank you!

**Comment by yuanjunchai** [comment]
Again, if you would like to just test, you'd better follow settings in the training domain. More details about how to make blurred data as test dataset has implemented in `codes/scripts/generate_mod_LR_bic.py`.
Like:
`        prepro = util.SRMDPreprocessing(up_scale, pca_matrix, random=False, para_input=10, kernel=21, noise=False,
                                        cuda=True, sig=sig, sig_min=0.2, sig_max=4.0, rate_iso=1.0, scaling=3,
                                        rate_cln=0.2, noise_high=0.0) #random(sig_min, sig_max) | stable kernel(sig)
        LR_img, ker_map = prepro(img_HR.view(1, C, H, W))`

**Comment by Ethean** [comment]
Hi, @yuanjunchai What should I do if I want to test a noisy image ?  Is it something like using 

> codes/scripts/generate_mod_LR_bic.py
>
to generate the LR images with changing the parameter noise to True and set the noise_high to 15 ? Like

> prepro = util.SRMDPreprocessing(up_scale, pca_matrix, random=False, para_input=10, kernel=21, noise=True, cuda=True, sig=sig, sig_min=0.2, sig_max=4.0, rate_iso=1.0, scaling=3, rate_cln=0.2, noise_high=15.0) #random(sig_min, sig_max) | stable kernel(sig) LR_img, ker_map = prepro(img_HR.view(1, C, H, W))
>
and retrain a new model? If I have to do so, how many epochs should I train this model? Thanks in advance!

**Comment by yuanjunchai** [comment]
Yes, you could retrain a new model for adding both noise and blur kernel. You could set 500k iteration and observe the convergence curve to make sure epochs about a new model.

**Comment by Ethean** [comment]
Thanks! I really appreciate your help! So if I take div2k and flickr2k as the training dataset, it would be enough as I set the epoch to 2500.  Now I am on it. Again, Thank you very much !

**Comment by yuanjunchai** [comment]
My pleasure :)

**Comment by Ethean** [comment]
Hi, @yuanjunchai  When I used code in
`codes/scripts/generate_mod_LR_bic.py`,
 I found it returned a shape of 1*11 kernel embedding, which was described in as h in the paper. It concatenated the noisy level with the original embedding. But in the paper author didn't claim that. So I changed the `call()` method  of class `SRMDPreprocessing()`  in  `codes/utils/util.py` .
The original code is 

> 
    def __call__(self, hr_tensor, kernel=False):
        ### hr_tensor is tensor, not cuda tensor
        B, C, H, W = hr_tensor.size()
        b_kernels = Variable(self.kernel_gen(self.random, B, tensor=True)).cuda() if self.cuda else Variable(self.kernel_gen(self.random, B, tensor=True))
        # blur
        if self.cuda:
            hr_blured_var = self.blur(Variable(hr_tensor).cuda(), b_kernels)
        else:
            hr_blured_var = self.blur(Variable(hr_tensor), b_kernels)
        # kernel encode
        kernel_code = self.encoder(b_kernels) # B x self.para_input
        # Down sample
        if self.cuda:
            lr_blured_t = b_GPUVar_Bicubic(hr_blured_var, self.scale)
        else:
            lr_blured_t = b_CPUVar_Bicubic(hr_blured_var, self.scale)

        # Noisy
        if self.noise:
            Noise_level = torch.FloatTensor(random_batch_noise(B, self.noise_high, self.rate_cln))
            lr_noised_t = b_GaussianNoising(lr_blured_t, Noise_level)
        else:
            Noise_level = torch.zeros((B, 1))
            lr_noised_t = lr_blured_t

        if self.cuda:
            Noise_level = Variable(Noise_level).cuda()
            re_code = torch.cat([kernel_code, Noise_level * 10], dim=1) if self.noise else kernel_code
            lr_re = Variable(lr_noised_t).cuda()
        else:
            Noise_level = Variable(Noise_level)
            re_code = torch.cat([kernel_code, Noise_level * 10], dim=1) if self.noise else kernel_code
            lr_re = Variable(lr_noised_t)
        return (lr_re, re_code, b_kernels) if kernel else (lr_re, re_code)
>
I changed the last few lines to return a 1*10 kernel embedding
> 
    def __call__(self, hr_tensor, kernel=False):
        ### hr_tensor is tensor, not cuda tensor
        B, C, H, W = hr_tensor.size()
        b_kernels = Variable(self.kernel_gen(self.random, B, tensor=True)).cuda() if self.cuda else Variable(self.kernel_gen(self.random, B, tensor=True))
        # blur
        if self.cuda:
            hr_blured_var = self.blur(Variable(hr_tensor).cuda(), b_kernels)
        else:
            hr_blured_var = self.blur(Variable(hr_tensor), b_kernels)
        # kernel encode
        kernel_code = self.encoder(b_kernels) # B x self.para_input
        # Down sample
        if self.cuda:
            lr_blured_t = b_GPUVar_Bicubic(hr_blured_var, self.scale)
        else:
            lr_blured_t = b_CPUVar_Bicubic(hr_blured_var, self.scale)

        # Noisy
        if self.noise:
            Noise_level = torch.FloatTensor(random_batch_noise(B, self.noise_high, self.rate_cln))
            lr_noised_t = b_GaussianNoising(lr_blured_t, Noise_level)
        else:
            Noise_level = torch.zeros((B, 1))
            lr_noised_t = lr_blured_t

        if self.cuda:
            #Noise_level = Variable(Noise_level).cuda()
            #re_code = torch.cat([kernel_code, Noise_level * 10], dim=1) if self.noise else kernel_code
            lr_re = Variable(lr_noised_t).cuda()
        else:
            #Noise_level = Variable(Noise_level)
            #re_code = torch.cat([kernel_code, Noise_level * 10], dim=1) if self.noise else kernel_code
            lr_re = Variable(lr_noised_t)
        return (lr_re, kernel_code, b_kernels) if kernel else (lr_re, kernel_code)
>
When I finished the image processing, I found that some image was noise-free. 
normal one 
![sig2 6_000062](https://user-images.githubusercontent.com/46132747/70019899-e7356c00-15c5-11ea-931a-a00efbfafce5.png)
abnormal one
![sig2 6_000076](https://user-images.githubusercontent.com/46132747/70019921-fcaa9600-15c5-11ea-9610-e6755f450b78.png)
 and I don't know why it comes to this.  Is this a training trick?

**Comment by yuanjunchai** [comment]
How about your setting, like 'rate_cln' or 'noise_high'??

**Comment by nanmehta** [comment]
@yuanjunchai 
Hello sir . I am new to this field of BLIND-SR, can you please help me out in understanding how have we wriitn the code for generating blur kernel

**Comment by Ethean** [comment]
@nan-rock see code in /codes/utils/util.py

**Comment by nanmehta** [comment]
Thank you sir. But I am not able to get the difference between stable and random codes of  isotropic, anisotropic gaussian ,batch kernels. Could u please help me out
